qiskit_machine_learning/neural_networks/circuit_qnn.py

```diff
@@ -147,7 +137,7 @@ def _compute_output_shape(self, interpret, output_shape, sampling) -> Tuple[int,
 
         # this definition is required by mypy
         output_shape_: Tuple[int, ...] = (-1,)
-        if sampling:
+        if sampling and self._quantum_instance is not None:
```

Comment:
The `_compute_output_shape` is called in constructor and if a valid quantum instance is given then the this code is run - the resultant shape is then passed to super constructor.

Now if QuantumInstance is None this code will not be executed. Should we be executing this later and somehow setting/updating the base class that stores output shape once a valid quantum instance is given via its setter.

Reply:
I don't quite follow... do you mean we should call `_compute_output_shape` within the setter (in addition to within the constructor?)

Reply:
There is an aspect  of `_compute_output_shape` that requires a QuantumInstance. In the constructor if that is None, and the method is called from there, the output shape it returns, if it should have gone via the QuantumInstance is not None path it will not. Hence once we do have a valid QuantumInstance I was thinking that method should be called and output_shape in the base class updated. Since passing in valid quantum instance is done later via the setter then it would be called from there - and I think we should do it even if there was a previous valid QuantumInstance since shots, which is used to compute output shape, may have been changed.

I think calling the compute_output_shape from both places is ok. In the constructor the output_shape gets passed up via the super.init, and if sampling and no QuantInst then it will be whatever the default is returned from that method which hopefully satisfies the base NeuralNetwork class that ultimately stores it. Later when the QI is updated, we can recompute the output shape from the setter, and update the base class. Since there is no setter we probably need to update _output_shape of the base class directly - though I note when going through the constructor, there is some sort of validity check too. Maybe thats not needed - if we think it is then possible make a private method to validate/set the output shape, that the constructor calls, and that we can call from a sub-class, in this case in the QI setter to update the shape.

Does that help?

Reply:
Ok thanks, so to make sure I understand:

We add a new function to the `NeuralNetwork` class, which handles the validation of the output shape:
```python
    def _validate_output_shape(self, output_shape):
        if isinstance(output_shape, int):
            output_shape = (output_shape,)
        if not np.all([s > 0 for s in output_shape]):
            raise QiskitMachineLearningError(
                f"Invalid output shape, all components must be > 0, but got: {output_shape}."
            )
```
 and within the constructor of `NeuralNetwork`, we just replace that validation with a call to this function. Within `CircuitQNN`, we add the following at the end of the QI setter:
 ```python
         output_shape = self._compute_output_shape(self._interpret, self._output_shape, self._sampling)
        self._validate_output_shape(output_shape)
        self._output_shape = output_shape
 ```
 is this on the right track?

Reply:
Seems like thats the right direction. Maybe the method should return output shape so its that which is set into the instance. Of course that only happens when its an `int` but it will allow the base class constuctor and you to use it the same way as seems your intent.